File: Assets/Scripts/Entity/XObject.cs
using System.Collections.Generic;


/// <summary>
/// 三个功能：
/// 1.对象生命周期
/// 2.分发事件
/// </summary>
public class XObject
{
   
    private Dictionary<XEventDefine, EventHandler> _eventMap;

    public bool Deprecated { get; set; }

    protected void Initilize()
    {
        Deprecated = false;
        EventSubscribe();
    }

    protected void Unload()
    {
        EventUnsubscribe();
        Deprecated = true;
    }


    /// <summary>
    /// 注册事件
    /// </summary>
    protected virtual void EventSubscribe()
    {
    }

    /// <summary>
    /// 销毁事件
    /// </summary>
    protected void EventUnsubscribe()
    {
        if (_eventMap != null)
            _eventMap.Clear();
        _eventMap = null;
        XEventMgr.singleton.RemoveRegist(this);
    }

    protected void RegisterEvent(XEventDefine eventID, XEventHandler handler)
    {
        if (!Deprecated)
        {
            if (_eventMap == null)
                _eventMap = new Dictionary<XEventDefine, EventHandler>();
            if (_eventMap.ContainsKey(eventID)) return;
            EventHandler eh = new EventHandler();
            eh.eventDefine = eventID;
            eh.handler = handler;
            _eventMap.Add(eventID, eh);
            XEventMgr.singleton.AddRegist(eventID, this);
        }
    }

    public virtual bool DispatchEvent(XEventArgs e)
    {
        if (!Deprecated && _eventMap != null)
        {
            if (_eventMap.ContainsKey(e.ArgsDefine))
            {
                var etor = _eventMap.GetEnumerator();
                while (etor.MoveNext())
                {
                    EventHandler eh = etor.Current.Value;
                    if (eh.eventDefine == e.ArgsDefine)
                    {
                        XEventHandler func = eh.handler;
                        if (func != null) func(e);
                        return true;
                    }
                }
            }
        }
        return false;
    }

    
}

File: tools_proj/XCPP/GameCore/XObject.cpp
#include "XObject.h"

XObject::XObject() {}

XObject::~XObject() {}

bool XObject::Initilize()
{
	EventSubscribe();
	return true;
}

void XObject::Uninitilize() 
{
	EventUnsubscribe();
}

void XObject::OnCreated() {}

void XObject::OnEnterScene() {}

void XObject::OnSceneReady() {}

void XObject::OnLeaveScene() {}

void XObject::Unload() {}

void XObject::EventSubscribe() {}

void XObject::EventUnsubscribe() {}

void XObject::RegisterEvent(XEventDefine eventID, XDelegate* handler)
{
	std::unordered_map<uint, XDelegate*>::iterator itr = eventMap.find((uint)eventID);
	if (itr == eventMap.end())
	{
		eventMap.insert(std::make_pair(eventID, handler));
		XEventMgr::Instance()->AddRegist(eventID, this);
	}
}

bool XObject::DispatchEvent(XEventArgs* e)
{
	XEventDefine def = e->getEventDef();
	if (def<0 || def>XEvent_MAX) return false;
	std::unordered_map<uint, XDelegate*>::iterator itr = eventMap.find(def);
	if (itr != eventMap.end())
	{
		XDelegate* eh = itr->second;
		if (eh) (*eh)(e, NULL);
		return true;
	}
	return false;
}




